lib/ip.go

```diff
@@ -12,18 +16,237 @@ type IP uint32
+	if len(endIPRaw) == 0 {
+		return IP6RangeInt{}, errors.New("invalid range end IP")
+	}
+	copy(startIPByte[:], []byte(startIPRaw.To16()))
+	copy(endIPByte[:], []byte(endIPRaw.To16()))
+
+	startIPInt := new(big.Int)
+	endIPInt := new(big.Int)
+	startIPInt.SetBytes(startIPByte[:])
+	endIPInt.SetBytes(endIPByte[:])
+
+	// ensure valid range
+	if startIPInt.Cmp(endIPInt) > 0 {
+		return IP6RangeInt{}, fmt.Errorf("invalid range: %v > %v", startIP, endIP)
+	}
+	return IP6RangeInt{
+		startIP: startIPInt,
+		endIP:   endIPInt,
+	}, nil
+}
+
+// RandIP4Range returns a list of randomly generated IPv4 addresses within
+// the range specified by `startIP` and `endIP`.
+//
+// note: `EvalIP4` must be called before this function as this function assumes
+// `startIP` and `endIP` is a correct range.
+func RandIP4Range(iprange IP4Range, noBogon bool) (net.IP, error) {
+	tmp := iprange.endIP - iprange.startIP
+	if tmp == 0 {
+		tmpIP := [4]byte{0, 0, 0, 0}
+		binary.BigEndian.PutUint32(tmpIP[:], uint32(iprange.startIP))
+		return tmpIP[:], nil
+	}
+
+	// get random IP and adjust it to fit range.
+	randIP := binary.BigEndian.Uint32(RandIP4(noBogon).To4())
+	randIP %= (uint32(iprange.endIP) - uint32(iprange.startIP))
+	randIP += uint32(iprange.startIP)
+	randIPbyte := [4]byte{0, 0, 0, 0}
+	binary.BigEndian.PutUint32(randIPbyte[:], randIP)
+	return net.IP(randIPbyte[:]), nil
+}
+
+// RandIP4ListWrite prints a list of new randomly generated IPv4 addresses.
+func RandIP4ListWrite(n int, noBogon bool) {
+	for i := 0; i < n; i++ {
+		fmt.Println(RandIP4(noBogon))
+	}
+}
+
+// RandIP4ListWrite prints a list of new randomly generated IPv4 addresses
+// within starting and IPs ending range.
+func RandIP4RangeListWrite(startIP, endIP string, n int, noBogon bool) error {
+	ipRange, err := NewIP4Range(startIP, endIP)
+	if err != nil {
+		return err
+	}
+	for i := 0; i < n; i++ {
+		ip, err := RandIP4Range(ipRange, noBogon)
+		if err != nil {
+			return err
+		}
+		fmt.Println(ip)
+	}
+	return nil
+}
+
 // IPFromStdIP returns a new IPv4 address representation from the standard
 // library's IP representation.
 func IPFromStdIP(ip net.IP) IP {
 	return IP(binary.BigEndian.Uint32(ip.To4()))
 }
 
+// RandIP6 returns a new randomly generated IPv6 address.
+func RandIP6(noBogon bool) net.IP {
+	ip := [16]byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
+IP:
+	binary.BigEndian.PutUint64(ip[0:], rand.Uint64())
+	binary.BigEndian.PutUint64(ip[8:], rand.Uint64())
+	var ip6 U128
+	ip6.Hi = binary.BigEndian.Uint64(ip[0:])
+	ip6.Lo = binary.BigEndian.Uint64(ip[8:])
+	if noBogon && IsBogonIP6(ip6) {
+		goto IP
+	}
+	return net.IP(ip[:])
+}
+
+// RandIP6Range returns a list of randomly generated IPv6 addresses within
+// the range specified by `startIP` and `endIP`.
+//
+// note: `EvalIP6` must be called before this function as this function assumes
+// `startIP` and `endIP` is a correct range.
+func RandIP6Range(ipRange IP6RangeInt, noBogon bool) (net.IP, error) {
```

Comment:
This function never actually returns a non-nil error, so we should only return a `net.IP`.